# Ship Detection from Satellite Imagery Testing

In [8]:
import numpy as np
import numpy.random as npr
import matplotlib.pyplot as plt
%matplotlib inline
plt.style.use('bmh')

X = np.load('ships_dataset/ship_data.npy')
t = np.load('ships_dataset/ship_labels.npy')

label_names = ['no_ship', 'ship']

from sklearn.model_selection import train_test_split

X_train, X_test, t_train, t_test = train_test_split(X, t, test_size = 0.2, random_state = 91)
print(X_test.shape)
X_flat = X_test.reshape(X_test.shape[0], -1)
X_flat.shape

(800, 80, 80, 3)


(800, 19200)

## Performance & Inference Metrics on the Test Set

In [9]:
import joblib
import time
from sklearn.metrics import f1_score
import numpy as np

def evaluate_model(path, X, y):
    model = joblib.load(path)

    start = time.time()
    preds = model.predict(X)
    end = time.time()

    total_time = end - start
    avg_time = total_time / len(X)

    f1 = f1_score(y, preds)

    return f1, total_time, avg_time


model_paths = {
    "logIso.pkl",
    "logLLE.pkl",
    "logPCA.pkl",
    "svmIso.pkl",
    "svmLLE.pkl",
    "svmPCA.pkl",
    "logregModel.pkl",  
    "svmModel.pkl",   
}

for path in model_paths:
    f1, total_time, avg_time = evaluate_model(path, X_flat, t_test)
    print(f"Model: {path}")
    print(f"  F1 Score: {f1:.4f}")
    print(f"  Total inference time: {total_time:.4f} sec")
    print(f"  Avg time per sample: {avg_time:.6f} sec\n")

Model: svmModel.pkl
  F1 Score: 0.7677
  Total inference time: 3.6666 sec
  Avg time per sample: 0.004583 sec

Model: svmLLE.pkl
  F1 Score: 0.0000
  Total inference time: 2.0357 sec
  Avg time per sample: 0.002545 sec

Model: logregModel.pkl
  F1 Score: 0.7990
  Total inference time: 0.0566 sec
  Avg time per sample: 0.000071 sec

Model: svmIso.pkl
  F1 Score: 0.8324
  Total inference time: 1.7153 sec
  Avg time per sample: 0.002144 sec

Model: logIso.pkl
  F1 Score: 0.8404
  Total inference time: 1.6647 sec
  Avg time per sample: 0.002081 sec

Model: logLLE.pkl
  F1 Score: 0.0000
  Total inference time: 1.8648 sec
  Avg time per sample: 0.002331 sec

Model: svmPCA.pkl
  F1 Score: 0.8495
  Total inference time: 0.1613 sec
  Avg time per sample: 0.000202 sec

Model: logPCA.pkl
  F1 Score: 0.8333
  Total inference time: 0.1485 sec
  Avg time per sample: 0.000186 sec



| Model Description                               | F1 Score | Total Inference Time (s) | Avg Time per Sample (s) |
| ----------------------------------------------- | -------- | ------------------------ | ----------------------- |
| **Support Vector Machine (SVM)**                | 0.7677   | 3.6666                   | 0.004583                |
| **SVM with LLE (Locally Linear Embedding)**     | 0.0000   | 2.0357                   | 0.002545                |
| **Logistic Regression**                         | 0.7990   | 0.0566                   | 0.000071                |
| **SVM with Isomap**                             | 0.8324   | 1.7153                   | 0.002144                |
| **Logistic Regression with Isomap**             | 0.8404   | 1.6647                   | 0.002081                |
| **Logistic Regression with LLE**                | 0.0000   | 1.8648                   | 0.002331                |
| **SVM with PCA (Principal Component Analysis)** | 0.8495   | 0.1613                   | 0.000202                |
| **Logistic Regression with PCA**                | 0.8333   | 0.1485                   | 0.000186                |
